src/main/kotlin/com/autonomousapps/model/Coordinates.kt

```diff
@@ -5,59 +5,111 @@ import dev.zacsweers.moshix.sealed.annotations.TypeLabel
 
 @JsonClass(generateAdapter = false, generator = "sealed:type")
 sealed class Coordinates(
-  open val identifier: String
+  open val identifier: String,
+  open val gradleVariantIdentification: GradleVariantIdentification
 ) : Comparable<Coordinates> {
 
   /**
    * [ProjectCoordinates] come before [ModuleCoordinates], which come before [IncludedBuildCoordinates], which come
    * before [FlatCoordinates].
    */
   override fun compareTo(other: Coordinates): Int {
-    return if (this is ProjectCoordinates) {
-      if (other is ProjectCoordinates) identifier.compareTo(other.identifier) else 1
-    } else if (this is FlatCoordinates) {
-      if (other is ProjectCoordinates) -1 else if (other is FlatCoordinates) gav().compareTo(other.gav()) else 1
-    } else {
-      when (other) {
-        is ProjectCoordinates -> -1
-        is FlatCoordinates -> -1
-        is ModuleCoordinates -> identifier.compareTo(other.identifier)
-        is IncludedBuildCoordinates -> identifier.compareTo(other.identifier)
+    return (
+      if (this is ProjectCoordinates) {
+        if (other is ProjectCoordinates) identifier.compareTo(other.identifier) else 1
+      } else if (this is FlatCoordinates) {
+        if (other is ProjectCoordinates) -1 else if (other is FlatCoordinates) gav().compareTo(other.gav()) else 1
+      } else {
+        when (other) {
+          is ProjectCoordinates -> -1
+          is FlatCoordinates -> -1
+          is ModuleCoordinates -> identifier.compareTo(other.identifier)
+          is IncludedBuildCoordinates -> identifier.compareTo(other.identifier)
+        }
+      }).let {
+        // after identifiers, compare capabilities
+        if (it == 0) {
+          gradleVariantIdentification.compareTo(other.gradleVariantIdentification)
+        } else
+          it
       }
-    }
   }
 
   /** Group-artifact-version (GAV) string representation, as used in Gradle dependency declarations. */
   abstract fun gav(): String
 
   fun toFileName() = gav().replace(":", "__") + ".json"
 
+  /**
+   * In case of an 'ADD' advice, the GradleVariantIdentification is directly sourced from the selected node
+   * in the dependency graph. It's hard (or impossible with Gradle's current APIs) to find the exact declaration that
+   * let to selecting that node. If we could find that declaration, we could use it for the ADD advice.
+   * Right now, we use the details from the node in the Graph which may contain more capabilities as you need
+   * to declare. In particular, it also contains the 'default capability', which makes it conceptually equal to
+   * Coordinates without capability.
+   * In order to correctly reduce advices (e.g. merge a REMOVE and an ADD to a CHANGE), we need the same Coordinates
+   * on both. So this method should be used to 'minify' the GradleVariantIdentification for ADD advices.
+   *
+   * @return A copy of this Coordinates without the 'default capability'
+   */
+  fun withoutDefaultCapability(): Coordinates {
```

Comment:
With all the different solutions I considered, I always tried to reduce assumptions and guesswork on top of the information we get out of Gradle. This method is basically what is left and for which I don't know if and how it could be further improved.
This is what the user gets in case of an ADD advice where we do not have an existing declaration to "reuse". I think this is a good start and if users report cases with unexpected advices, this can be tuned further if we have real-world reproducers.